Adding namespace to SAN

davidhadas opened the pull request:
Fixes #264 

This PR aims at a minimalistic change 
1. Add namespace SANs to certificates + modify tests accordingly. 
2. Since it is meant to be used also for mTLS, the produced certificates need to serve dual use  - both as a client certificate and as a server certificate. For example, the Activator will use this certificate as both client and server. This resulted in merging the creation of a client template with that of a server template. 

Note:
While working on this PR many other opportunities to merge code poped up. The terminology used by the code (data-plane, control-plane) is somewhat unclear and it is hard to see the relationship between the naming and the certificates created and their use. These are left for a separate PR after a discussion of the correct terminology and the intent for the separation between control-plane and data-plane certificates.  


Items for review:
 - Do we want the SAN to be \<namespace\> or \<prefix\>-\<namespace\>? For some reason, the author of this PR followed knative-\<namespace\> but cant justify why this prefix is helpful.

davidhadas commented:
/retest

davidhadas commented:
Current code terminology is a bit confusing.
Is there a clear distinction between "control-plane" certificates "data-plane" certificates? 
Why are they separated in code? 
Does it really matter from TLS and certificate point of view if the entity (client or server) is classified as a control or data? 

I would like to first have a fix for SANs in this PR, then consider if we need to open an issue and create a second PR to rearrange the code by removing the current separation of "control-plane" vs.  "data-plane" certificates if this is not needed. Alternatively, we may need to add some remarks and make it more clear or improve naming in this project to make it clearer when we refer to clients or server etc..

davidhadas commented:
/retest

davidhadas commented:
/cc @ReToCode @nak3 @evankanderson

nak3 commented on pkg/certificates/certs.go:
```diff
@@ -48,14 +49,14 @@ func createCertTemplate(expirationInterval time.Duration) (*x509.Certificate, er
 // Create cert template suitable for CA and hence signing
 func createCACertTemplate(expirationInterval time.Duration) (*x509.Certificate, error) {
-	rootCert, err := createCertTemplate(expirationInterval)
+	rootCert, err := createCertTemplate(expirationInterval, FakeDnsName)
```
Will it create duplicated `FakeDnsName` in `DNSNames: []string{FakeDnsName, FakeDnsName}` in `createCertTemplate()`?

davidhadas replied:
Yes it will. This is at the CA Template where I am not sure why we have Subject Alternate Name values in the first place. I believe it can probably be removed in a later cleanup PR.

nak3 commented on pkg/certificates/certs.go:
```diff
@@ -70,36 +71,20 @@ func createCACertTemplate(expirationInterval time.Duration) (*x509.Certificate,
-	clientCert.ExtKeyUsage = []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth}
-	clientCert.Subject = pkix.Name{
+	cert.KeyUsage = x509.KeyUsageDigitalSignature
+	cert.ExtKeyUsage = []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth}
```
Sorry I am also not sure why the original author separated the data-plane vs control-plane as you raised the question https://github.com/knative-sandbox/control-protocol/pull/265#issuecomment-1478185451
But if it is no problem for having both `x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth` as security practice, I think it is not necessary for the separation.

evankanderson replied:
... thinking about this, I can't see any _special_ risk in a having a certificate with both usages.

My thinking:

* Key Usages instruct the other side (client or server) about whether the cert is intended for that authorization purpose.
* In order to confuse the client / server usage, you would need to do something like initiate a simultaneous TCP handshake (where the victim _thinks_ they are the client, but the remote side _thinks_ they are the server).  This is likely to fall down pretty quickly after the TLS handshake, if not before, because none of our protocols are "palindromic" (you can speak them the same as client or server).
* Additionally, we're only using this for network communication, not for storing records, so it would be hard to "trick" a server into signing something and then putting it in storage on the server's behalf.

ReToCode commented on pkg/certificates/certs.go:
```diff
@@ -170,31 +155,31 @@ func CreateControlPlaneCert(ctx context.Context, caKey *rsa.PrivateKey, caCertif
 		return nil, err
 	}
-	serverCertTemplate, err := createServerCertTemplate(expirationInterval)
+	certTemplate, err := createClientServerCertTemplate(expirationInterval, san)
```
Naming is really a bit off overall. Why is this ClientServer? The certs are just used as TLS server certs, right? If we at some point add a client cert for mTLS, we should start talking about client-certs.

evankanderson replied:
👍  that a single cert for both client and server seems safe (see above, and usage elsewhere with, for example, service meshes).

I'd recommend fixing the _public_ naming now along with the breaking API change, and leaving the internal function naming to sort out later.

ReToCode commented on pkg/certificates/certs.go:
```diff
@@ -70,36 +71,20 @@ func createCACertTemplate(expirationInterval time.Duration) (*x509.Certificate,
 		Organization: []string{Organization},
 		// Do not use the same value with CA.
-		CommonName: "control-plane-client",
+		CommonName: "control-plane-client-server",
```
Same here, why is this client-server?

davidhadas replied:
See above

ReToCode reviewed (changes requested):
I agree with you wrt naming. Maybe we could also think about something better for the CA:
`CommonName: "control-plane"`
I think its not ideal, as this is also used to sign data-plane certs and potentially more in the future. Could it be something like `knative-internal-ca`?

ReToCode reviewed (approved):
/lgtm

evankanderson commented on pkg/certificates/certs.go:
```diff
@@ -70,36 +71,20 @@ func createCACertTemplate(expirationInterval time.Duration) (*x509.Certificate,
 		Organization: []string{Organization},
 		// Do not use the same value with CA.
-		CommonName: "control-plane-client",
+		CommonName: "control-plane-client-server",
```
Should we put in the SAN for the CommonName?

If not (and we're not worried about changing the CN), I might remove `control-plane` from the CN, but this is also fine.

evankanderson replied:
This is really a nit, and isn't super-important either way.

evankanderson commented on pkg/certificates/certs.go:
```diff
@@ -142,7 +127,7 @@ func CreateCACerts(ctx context.Context, expirationInterval time.Duration) (*KeyP
 
 // CreateControlPlaneCert generates the certificate for the client
-func CreateControlPlaneCert(ctx context.Context, caKey *rsa.PrivateKey, caCertificate *x509.Certificate, expirationInterval time.Duration) (*KeyPair, error) {
+func CreateControlPlaneCert(ctx context.Context, caKey *rsa.PrivateKey, caCertificate *x509.Certificate, expirationInterval time.Duration, san string) (*KeyPair, error) {
```
I'd suggest not making a breaking change in the library.  Two options:

1. Make this `san ...string`, and have empty string maintain the current behavior.
2. Add a _new_ function, e.g. `CreateServerCert` which takes the new arguments.

When we make breaking changes in these libraries, it gums up the dependency / buoy process for upgrading all the clients.

We could still remove this function (if we want) after we've migrated the callers, but I'd prefer not to get in the habit of making breaking library changes in broad dependencies (and, sadly, `control-protocol` has broad dependencies).

davidhadas replied:
nice. 
Let's g with option 1

evankanderson commented on pkg/certificates/certs.go:
```diff
@@ -170,31 +155,31 @@ func CreateControlPlaneCert(ctx context.Context, caKey *rsa.PrivateKey, caCertif
 
 // CreateDataPlaneCert generates the certificate for the server
-func CreateDataPlaneCert(ctx context.Context, caKey *rsa.PrivateKey, caCertificate *x509.Certificate, expirationInterval time.Duration) (*KeyPair, error) {
+func CreateDataPlaneCert(ctx context.Context, caKey *rsa.PrivateKey, caCertificate *x509.Certificate, expirationInterval time.Duration, san string) (*KeyPair, error) {
```
Ditto in terms of breaking signatures.

If `CreateControlPlaneCert` and `CreateDataPlaneCert` do the same thing now, it seems like a new function might be for the best long-run.

davidhadas replied:
Let's leave this for a subsequent PR. 
There are some opportunities to merge code and rethink what we name here as data-plane or control-plane and if it has anything to do with data or control. 
We also need to think if we need separation of data and control certificate wise...
If so, we don't do that today (considering that everything in knative-serving now uses the same certificate ;) )

davidhadas replied:
Now realizing that this relates to the comment below asking to fix now Public Naming...

evankanderson commented on pkg/network/tls.go:
```diff
@@ -33,7 +33,7 @@ func LoadServerTLSConfigFromFile() (*tls.Config, error) {
 		ClientCAs:    certPool,
 		ClientAuth:   tls.RequireAndVerifyClientCert,
-		ServerName:   certificates.FakeDnsName,
+		ServerName:   "knative-myns",
```
Should we check the old name _as well as_ the new name?

davidhadas replied:
My assumption is that we are on the path to getting to namespace SANs, we keep the old names just to ensure we get there in gradual non disruptive steps...

evankanderson reviewed (commented):
Overall, I like this, but I think we might want to adjust things a little bit:

* Bite the bullet and add a new `CreateServerCert` method which takes a _list_ of SANs.
* Expose `fakeDnsName` as `LegacyFakeDnsName`
* Migrate callers to `CreateServerCert`, and add `LegacyFakeDnsName` to the list of SANs provided at the caller site.
* (Add a flag at the caller site to be able to drop `LegacyFakeDnsName` in the future)

davidhadas commented on pkg/certificates/certs.go:
```diff
@@ -36,7 +36,8 @@ var randReader = rand.Reader
 
 // Copy-pasted from https://github.com/knative/pkg/blob/975a1cf9e4470b26ce54d9cc628dbd50716b6b95/webhook/certificates/resources/certs.go
-func createCertTemplate(expirationInterval time.Duration) (*x509.Certificate, error) {
+func createCertTemplate(expirationInterval time.Duration, san string) (*x509.Certificate, error) {
```
sure

davidhadas commented on .codecov.yaml:
I added it since it was required to get this PR pass the tests...

davidhadas commented:
> * Bite the bullet and add a new `CreateServerCert` method which takes a _list_ of SANs.
Added `CreateCert` as it creates cert for both client and server

> * Expose `fakeDnsName` as `LegacyFakeDnsName`
Already exposed as `FakeDnsName`, changed to `LegacyFakeDnsName`

> * Migrate callers to `CreateServerCert`, and add `LegacyFakeDnsName` to the list of SANs provided at the caller site.
Migrated all internal callers, we will need to now migrate any other caller from other packages. 

> * (Add a flag at the caller site to be able to drop `LegacyFakeDnsName` in the future)
I think a PR will do it.

davidhadas commented:
Do we want the SAN to be <namespace> or <prefix>-<namespace>? 
For some reason, the author of this PR followed knative-<namespace> but can't justify why this prefix is helpful...
Should we leave it? or change it?

evankanderson commented:
> Do we want the SAN to be <namespace> or <prefix>-<namespace>? For some reason, the author of this PR followed knative-<namespace> but cant justify why this prefix is helpful.

I think the prefix is helpful for helping users connect the dots if the prefix is sufficiently specific.

evankanderson reviewed (commented):
/lgtm
/approve

dprotaso commented on pkg/certificates/constants.go:
```diff
@@ -17,8 +17,9 @@ limitations under the License.
 const (
-	Organization = "knative.dev"
-	FakeDnsName  = "data-plane." + Organization
+	Organization        = "knative.dev"
+	LegacyFakeDnsName   = "data-plane." + Organization
+	DataPlaneNamePrefix = "knative-"
```
A lot of downstream repos seem to use the `FakeDnsName` constant - so this is breaking? Are you doing follow up PRs to the various net-* repos?

davidhadas replied:
I suggest we fix this here  - see https://github.com/knative-sandbox/control-protocol/pull/269
And later follow up with a more gradual change in dependencies.

davidhadas commented:
I suggest we fix this here  - see https://github.com/knative-sandbox/control-protocol/pull/269
And later follow up with a more gradual change in dependencies.
